serialization/serialization_v2.py

```diff
@@ -0,0 +1,1032 @@
+
+        # Any of the following could potentially cause further recursive calls,
+        # thus check the recursive flag at this point.
+        if recursive:
+            assert not self._inlining_stage  # should not get here when inlining
+            raise _SerializationDependsOnNotYetSerializedOtherVarException(
+                _AssignQueueItem(value=value, suggested_var_name=prefix),
+            )
+        assert name
+
+        if isinstance(value, dict):
+            return self._serialize_dict(value, prefix)
+        if isinstance(value, list):
+            return self._serialize_list(value, prefix)
+        if isinstance(value, tuple):
+            return self._serialize_tuple(value, prefix)
+        if isinstance(value, set):
+            return self._serialize_set(value, prefix)
+        if isinstance(value, functools.partial):
+            return self._serialize_functools_partial(value, name)
+        if isinstance(value, CodeFromFunction):
+            return self._serialize_code_from_function(value, name)
+        if isinstance(value, (CodeFromFile, ExplicitHash, ExternalImport, NonhashedCode)):
+            raise ValueError(
+                f"Cannot serialize {type(value).__name__} in config dict. "
+                "It does not represent a value and should go into python_prolog/python_epilog."
+            )
+        if isinstance(value, ModuleType):
+            return self._serialize_module(value, name)
+        if isinstance(value, MethodType):
+            return self._serialize_method(value, name)
+
+        # special cased custom types
+        if _isinstance_returnn_dim(value):
+            return self._serialize_dim(value, prefix)
+        if _isinstance_returnn_cached_file(value):
+            return self._serialize_cached_file(value, prefix)
+        if isinstance(value, Checkpoint):
+            return self._serialize_via_repr(value, need_brackets_when_inlined=False)
+        if isinstance(value, CodeWrapper):
+            return self._serialize_via_repr(value, need_brackets_when_inlined=True)
+
+        if isinstance(value, (type, FunctionType, BuiltinFunctionType)) or (
+            getattr(value, "__module__", None) and getattr(value, "__qualname__", None)
+        ):
+            return self._serialize_global(value=value, name=name)
+
+        # Generic fallback using __reduce__ or __reduce_ex__.
+        return self._serialize_reduce(value, name)
+
+    def _serialize_reduce(self, value: Any, name: str) -> Union[_PyEvalCode, _PyCodeWithDeferredStateQueueItem]:
+        # Generic fallback using __reduce__ or __reduce_ex__.
+        # This is very much following the original pickle logic (slightly simplified).
+        value_ref = _Ref(value)
+        if value_ref in self.reduce_cache_by_value_ref:
+            rv = self.reduce_cache_by_value_ref[value_ref]
+        else:
+            reduce = getattr(value, "__reduce_ex__", None)
+            reduce_proto = 4  # not sure...
```

Comment:
> ` # not sure...`

why not? can we make sure?

Reply:
Yea, sounds good to change the comment like that.